src/main/java/hudson/plugins/sloccount/SloccountBuildAction.java

```diff
@@ -46,15 +45,16 @@ public String getUrlName() {
     
     @Override  
     public Collection<? extends Action> getProjectActions() {
-        return this.projectActions;
+        buildProjectActions();
+        return Collections.unmodifiableList(this.projectActions);
     }
 
     private void buildProjectActions() {
         if (this.projectActions != null) {
             return;
         }
 
-        this.projectActions = Collections.synchronizedList(new ArrayList<SloccountProjectAction>());
+        this.projectActions = new CopyOnWriteArrayList<>();
         this.projectActions.add(new SloccountProjectAction(this.build.getParent(), this.numBuildsInGraph));
```

Comment:
This code likely needs to be synchronized or reworked to 

```java
CopyOnWriteArrayList<SloccountProjectAction> actions = new CopyOnWriteArrayList<>();
actions.add(new SloccountProjectAction(this.build.getParent(), this.numBuildsInGraph));
this.projectActions = actions;
```

Otherwise there is a risk of race conditions even in `CopyOnWriteArrayList` IIUC (two parallel `buildProjectActions()` calls may cause data duplication)

Reply:
Makes sense. I was wondering whether it wouldn't be better/cleaner to have the whole method synchronised.